pippy/PipelineSchedule.py

```diff
@@ -0,0 +1,676 @@
+                # add recv for the NEXT stage, do not do this for first stage
+                if bwd_stage_id_next is not None:
+                    ops = backward_stage_next.get_bwd_recv_ops()
+                    if mb_id_bwd != len(microbatches) - 1:
+                        requests.extend(ops)
+
+                if requests:
+                    logger.info(
+                        f"rank: {self.pp_id}, current stage_id {self.stage_id_to_global_stage_id[bwd_stage_id]}, next stage_id {self.stage_id_to_global_stage_id[bwd_stage_id_next]} requests - {[(req.op, req.peer) for req in requests]}"
+                    )
+                    backward_batched_op_handles.append(
+                        dist.batch_isend_irecv(requests).pop()
+                    )
+
+        logger.info("Step exiting")
+
+
+def main(**kwargs):
+    torch.manual_seed(42)
+    print(f"MY KWARGS ARE {kwargs}")
+    rank = kwargs["rank"]
+    local_rank = kwargs["local_rank"]
+    world_size = kwargs["world_size"]
+
+    setup(local_rank, world_size)
+    logger.info(
+        f"====== World Rank {rank}, Local Rank {local_rank}, World Size {world_size} main ======"
+    )
+
+    input_dim = 4000
+    hidden_dim = 8000
+    output_dim = 4000
+
+    module_list = torch.nn.ModuleList(
+        modules=[
+            MLP(input_dim, hidden_dim, output_dim) for i in range(world_size)
+        ]
+    )
+    microbatch_size = 8
+    global_batch_size = 64
+    assert global_batch_size % microbatch_size == 0
+    n_microbatches = int(global_batch_size / microbatch_size)
+    n_pp = world_size
+
+    x = torch.randn([microbatch_size, input_dim]).to("meta")
+
+    stage_model = PipelineStage(
+        module_list[rank], rank, world_size, rank, world_size, x
+    )
+    stage_model.init_p2p_neighbors()
+
+    stage_model_looped = [
+        PipelineStage(
+            module_list[rank],
+            stage_id=(world_size * i) + rank,
+            num_stages=world_size * world_size,
+            rank=rank,
+            world_size=world_size,
+            meta_input=x,
+        )
+        for i in range(world_size)
+    ]
+    x_cuda_empty = torch.empty_like(x, device="cuda")
+    microbatches = [
+        torch.randn_like(x_cuda_empty) for _ in range(n_microbatches)
+    ]
+
+    for schedule in kwargs["schedules"]:
+        logger.info(f"====== Rank {rank} running schedule {schedule} ======")
+        if schedule == "gpipe":
+            pipeline = PipelineScheduleGPipe(stage_model)
+        elif schedule == "looped_bfs":
+            pipeline = PipelineScheduleLoopedBFS(stage_model_looped)
+        elif schedule == "looped_dfs":
+            pipeline = PipelineScheduleLoopedDFS(
+                stage_model_looped,
+                n_microbatch=n_microbatches,
+                pp_id=rank,
+                n_pp=n_pp,
+            )
+
+        logger.info(f"====== Rank {rank} profile ======")
+
+        # with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]) as prof:
```

Comment:
future item - maybe we can switch this to a context manager that is null context when not profiling, to avoid having profiler hanging around when not in use.

Reply:
Makes sense, will update that.

Also hope splitting up the file is okay for the IBM folks to use, if not please direct them to this version of the single file

https://github.com/pytorch/PiPPy/blob/6c8163ae936792b8ce3698a71934e3ec0daaa9e4/pippy/PipelineSchedule.py